### ISSUE TITLE
改写Op，添加对index tensor的支持

### ISSUE THREAD

**Issue by silentspring2** [acada-sjtu/EdgeTraining#33]
改写以下Op的计算：
- [ ] Conv
- [ ] Dense
- [ ] element-wise add
- [ ] element-wise time
- [ ] batch normalization

改写以下Op，添加对一些稀疏系数的传递功能：
- [ ] ReLU
- [ ] MaxPool
- [ ] AvgPool

**Comment by silentspring2** [comment]
一个问题是Prune掉channel的feature map对BN是怎么处理的？是当作0算？还是不算？
如果是前者，那么过BN层就会失去稀疏性；如果是后者，那么怎么和训练小model做区分？

**Comment by ACA-Lab-SJTU** [comment]
现在一种做法是在稀疏化（压缩的时候）把BN层与量化结合在一起，融合进临近的网络层。
*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by silentspring2** [comment]
之前有过文章讲述BN和Conv等层的融合，但这貌似是graph level的optimzation，我们会研究其可行性。而我们当前的设想似乎是以op level实现。

BN本身由两个乘加组成，第一个乘加以每个channel为单独单位，因此对于pruned掉的channel不会有影响。XXQ正在survey第二个乘加，观测其对pruned掉的channel的影响。

**Comment by ACA-Lab-SJTU** [comment]
对没错，是graph level的，是否我们就不用考虑了呢？


-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by silentspring2** [comment]
对于BN的单独优化，已经放在[Issue 34](https://github.com/acada-sjtu/EdgeTraining/issues/34)中讨论。

**Comment by silentspring2** [comment]
Conv2d Sparse的相关工作，将放在[Issue 35](https://github.com/acada-sjtu/EdgeTraining/issues/35)中讨论。

**Comment by silentspring2** [comment]
对于Dense层的支持，可以借用topi.sparse.dense中CSR格式的方法实现。

Dense目前看来并不是我们的main contribution，主要是保证网络通路。

**Comment by silentspring2** [comment]
ReLU only needs support from NNVM level. A new op named ReLU_sparse is added on nnvm/src/top/nn/nn.cc.

**Comment by silentspring2** [comment]
@xuxiaoqiao @ACA-Lab-SJTU 
Please consider the difference between these two implementations:
> x_data, x_index = sym.conv2d_sparse(_some params_)
y_data, y_index = sym.relu_sparse(x_data, x_index)
z_data, z_index = sym.conv2d_sparse(y_data, y_index, _other params_)

And
> x_data, x_index = sym.conv2d_sparse(_some params_)
y_data = sym.relu(x_data)
z_data, z_index = sym.conv2d_sparse(y_data, x_index, _other params_)

Given that ReLU and Pooling Layers pass the index tensor and compute only on the data tensor, is it necessary for us to introduce some new Ops to deliver such functions?

Thanks

**Comment by silentspring2** [comment]
How are these two implementations affect forward and backward process?

**Comment by xuxiaoqiao** [comment]
我觉得两种没有实质性区别

**Comment by silentspring2** [comment]
After discussion, we don't think these two implementations vary from each other, given it is not likely to have any impact on both forward or backward process.

Hence, the sparse implementation of pool2d and relu is not necessary.

**Comment by silentspring2** [comment]
If @rogercmq 's experiments prove that double masking techniques can be applied on structured sparsity, these will be no need to implement BN with sparsity as well.

**Comment by ACA-Lab-SJTU** [comment]
I’m not sure whether the experiment can provide enough credential. Does that mean every DNN training on the edge using our framework should have to use double masking  technique around the BN layer? Is that a heavy constraint for our framework?
*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by silentspring2** [comment]
It is hard to answer the question because the answer really depends on the experiment. Besides, double masking numerically is equal to omitting the beta term in BN.

If double masking is not suitable, the inference/training cost is going to increase a lot. So the benefit is a lot, while the cost is unknown.

**Comment by ACA-Lab-SJTU** [comment]
Yes. double masking is equivalent to eliminating the beta term in BN. But, double masking is not our contribution. Right?
*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by ACA-Lab-SJTU** [comment]
By contribution, I mean our solution is already embedded in the tool. Algorithm develop doesn’t need to worry about it. 
How does the double masking paper prove the efficiency and correctness of their method in math? If they already have an elegant proof. Then, we only need to prove the numerical equivalence between these two approach. 
*******************************************
Li Jiang（蒋力）, Ph.D., Associate Professor
SEIEE-3-521, Department of Computer Science & Engineering,
Shanghai Jiaotong University, Shanghai, China
Email: jiangli@cs.sjtu.edu.cn <mailto:jiangli@cs.sjtu.edu.cn>
Web: www.cs.sjtu.edu <http://www.cs.sjtu.edu/>.cn/~jiangli

**Comment by silentspring2** [comment]
Yes, it is. Prof Yuan Xie has proven that double masking can be used for non-structured pruning, but we are not sure how it will affect the accuracy, mobile robustness or convergence speed when applied on structured pruning, more specifically, channel/filter pruning. 

If the effect is proven to be neglectable, then we can safely implement such techniques into our system and, very likely, save a lot of memory/execution time.

**Comment by silentspring2** [comment]
Double masking may not be the largest contribution of our work, but it is likely to save a lot of memory/execution time. 

I agree that proving the equivalence by math is also very useful when it comes to the reason we implement double masking in our system.
